src/dtree.ml

```diff
@@ -0,0 +1,200 @@
+    {!recfield:swap} indicates whether the columns of the matrix have been
+    swapped before the switch, with an int indicating the column which has
+    been swapped with the first one (or None if no swap is performed) and
+    {!recfield:children} contains the subtrees. *)
+and node_data = { switch : term option
+                ; swap : int option
+                ; children : t list}
+
+(** [iter l n f t] is a generic iterator on trees; with function [l] performed
+    on leaves, function [n] performed on nodes, [f] returned in case of
+    {!const:Fail} on tree [t]. *)
+let iter : (term option -> action -> 'a) ->
+  (term option -> int option -> t list -> 'a) ->
+  'a -> t -> 'a = fun do_leaf do_node fail t ->
+  let rec loop = function
+    | Leaf(teo, a)                                   -> do_leaf teo a
+    | Fail                                           -> fail
+    | Node({ switch = s ; swap = p ; children = c }) ->
+      do_node s p (List.map loop c) in
+  loop t
+
+(** [to_dot f t] creates a dot graphviz file [f].gv for tree [t]. *)
+let to_dot : string -> t -> unit = fun fname tree ->
+  let module F = Format in
+  let module P = Print in
+  let ochan = open_out (fname ^ ".gv") in
+  let ppf = F.formatter_of_out_channel ochan in
+  let nodecount = ref 0 in
+  F.fprintf ppf "graph {@[<v>" ;
+  let pp_opterm : term option pp = fun oc teo -> match teo with
+    | Some(t) -> P.pp oc t
+    | None    -> F.fprintf oc "d" in
+  let rec write_tree : int -> t -> unit = fun father_l tree ->
+  (* We cannot use iter since we need the father to be passed. *)
+    match tree with
+    | Leaf(t, a)  ->
+      incr nodecount ;
+      F.fprintf ppf "@ %d [label=\"" !nodecount ;
+      let _, acte = Bindlib.unmbind a in
+      P.pp ppf acte ; F.fprintf ppf "\"]" ;
+      F.fprintf ppf "@ %d -- %d [label=\"" father_l !nodecount ;
+      pp_opterm ppf t ; F.fprintf ppf "\"];"
+    | Node(ndata) ->
+      let { switch = t ; swap = swa ; children = c } = ndata in
+      incr nodecount ;
+      let tag = !nodecount in
+      F.fprintf ppf "@ %d [label=\"" tag ;
+      F.fprintf ppf "%d" (match swa with None -> 0 | Some(i) -> i) ;
+      F.fprintf ppf "\"]" ;
+      F.fprintf ppf "@ %d -- %d [label=\"" father_l tag ;
+      pp_opterm ppf t ; F.fprintf ppf "\"];" ;
+      List.iter (fun e -> write_tree tag e) c ;
+    | Fail        ->
+      incr nodecount ;
+      F.fprintf ppf "@ %d -- %d [label=\"%s\"];" father_l !nodecount "f"
+  in
+  write_tree 0 tree ;
+  F.fprintf ppf "@.}@\n@?" ;
+  close_out ochan
+
+(** [nth l i] gives {!const:Some}[(i)]th element of [l] if it exists, None
+    otherwise. *)
+let rec nth : 'a list -> int -> 'a option = fun l i ->
+  match l with
+  | []                 -> None
+  | x :: _  when i = 0 -> Some(x)
+  | _ :: xs when i > 0 -> nth xs (i - 1)
+  | _ (* For linter *) -> assert false
+
+(** [swapf x i] the [i]th element of [x] with the first one. *)
+let swapf : 'a list -> int -> 'a list = fun li ind ->
```

Comment:
Should be in extra.ml.

Reply:
What the "f" stands for?

Reply:
done in 8a424b0

Reply:
The "f" stands for first since we swap with the first element.  Clear enough or a verbose "swap_first" is preferred?

Reply:
Renamed to `swap_head` in f5681bf